R/cmapplot.R

```diff
@@ -22,23 +22,53 @@
 
   family <- name <- path <- NULL
 
-  # check for Whitney
-  all_fonts <- systemfonts::system_fonts()
-  whitney_core <- all_fonts$name[all_fonts$name %in% c("Whitney-Medium", "Whitney-Book", "Whitney-Semibold")]
-  assign("use_whitney", length(whitney_core) >= 3, envir = cmapplot_globals)
+  # if font registry already contains whitney core, set use_whitney == TRUE
+  check_for_whitney_core(set_global = TRUE)
+
+  # else, register all whitney fonts in systemfonts. Then, if font registry
+  # contains whitney core, set use_whitney == TRUE
+  #
+  # (This is necessary because R looks up font by "family" with only basic
+  # variation (bold, italic, etc), so fonts like "Whitney-Book" are inaccessible
+  # by default. Font registration allows us to use the font's "name" as it's
+  # "family" so R can identify it.)
+  if(!get("use_whitney", envir = cmapplot_globals)){
+
+    whitney_fonts <- systemfonts::system_fonts() %>%
+      dplyr::filter(family == "Whitney") %>%
+      dplyr::select(name, path)
 
-  if(get("use_whitney", envir = cmapplot_globals)){
-    # Register all Whitney fonts (note: this registers italic fonts both as
-    # variants of core fonts and as standalone fonts, so there is some
-    # duplication.)
-    whitney_fonts <- select(filter(all_fonts, family == "Whitney"), name, path)
     purrr::walk2(whitney_fonts$name, whitney_fonts$path, systemfonts::register_font)
 
-    # Update font variables
+    check_for_whitney_core(set_global = TRUE)
+  }
+
+  # else, if not on Windows, check for /Library/Fonts directory.
+  # register all Whitney fonts in this folder.
+  # If font registry contains whitney core, set use_whitney == TRUE
+  if(!get("use_whitney", envir = cmapplot_globals) & .Platform$OS.type != "windows"){
+
+    user_dir <- paste0(Sys.getenv("HOME"), "/Library/Fonts")
+
+    if(dir.exists(user_dir)){
+      whitney_fonts <- list.files(user_dir, full.names = TRUE) %>%
+        as.data.frame() %>%
+        rlang::set_names("path") %>%
+        dplyr::filter(stringr::str_detect(path, "Whitney")) %>% # Will error for username of "Whitney"
```

Comment:
To avoid the potential error for unluckily named users, can't the `str_detect()` check for "Whitney-" instead of "Whitney"?